### ISSUE TITLE
Shape keys and NLA channels not being exported

### ISSUE THREAD

**Issue by samvila** [godotengine/godot-blender-exporter#90]
Hi,

I'm providing a sample file that includes armature animation+shape keys animation+nla. The armature animation gets exported correctly but not the other two.

[sample.zip](https://github.com/godotengine/godot-blender-exporter/files/2350239/sample.zip)

**Comment by Jason0214** [comment]
Ah, the NLA track is not supported right now. (the current usage of NLA is to export muted nla track as separate animation look http://docs.godotengine.org/en/latest/getting_started/workflow/assets/escn_exporter/animation.html). However the given sample is a good example showing the usefulness of NLA track, I think I will implement it in these days.

Besides, your sample also makes me find a hiding bug, the multiple `nla_strip`s inside one `nla_track`  is exported as separate actions, so you will see something like this:

it is indeed a bug, but it also confirms shape key animation works correct, the problem is in the layout of animation tracks.

And there is a small issue in your sample, one mesh object `assistant_body` has an active animation `anim` which is a skeleton animation, this one actually contribute to nothing, if you open the `Dope sheet` work space, you can see blender give you a lot of warnings (also that's the why you get a lot of warning in exporting)

I am trying to find a work around for all these problems, but unfortunately can not come about one, sorry about that.

**Comment by samvila** [comment]
Hi @Jason0214 , that's a fantastic review about my report. Thanks! Looking forward to test the sample file in future versions of your script and see if it will be able to transfer everything correctly. One thing that I didn't get from your reply is this sentence:

>  but unfortunately can not come about one, sorry about that.

Could you elaborate this a little bit? Sorry didn't understand that part.

**Comment by Jason0214** [comment]
I just mean I know what you want to do, but it can not accomplished with current `master`,  so you have to wait for #91. sorry for the confusion.

**Comment by samvila** [comment]
Hi @Jason0214 , any idea about when your PR is going to be merged? It seems like you fix most of the problems but still not merged. Looks very promising :-)

**Comment by Jason0214** [comment]
It is kind of a big change, so I was planned to let some other developers take a look at it.  I did some minor update of that branch days ago, I am already satisfied with my fix, so I think I will do some final checking and try to merge it in two or three days.  Sorry for the long waiting.

**Comment by samvila** [comment]
Hi @Jason0214 , I gave a try to your commit #91 using my sample file and I still get errors, is that the expected behavior or is it something I'm doing wrong?

**Comment by Jason0214** [comment]
Sorry, it's my bad :(     It should be fixed now. By the way, you may want to use the 'Scene Animation' option for your blend to be better exported.

**Comment by samvila** [comment]
I think there's still some problem, when I try to export using the latest build I still get a lot of errors in Blender and doesn't generate the Godot scene file.

**Comment by Jason0214** [comment]
could you tell me about the blender version you are using, what exporting options you are using?  I really want to help, but I can not reproduce the error now.

**Comment by samvila** [comment]
I'm using Blender 2.79b and Godot 3.1 Alpha2, both in Windows 7 64 bit.

Using my sample file from the first post I get this errors:

![blender2godot](https://user-images.githubusercontent.com/8132267/47960597-64ee0900-dfbb-11e8-8442-ad8dd5a54c83.png)

This generates a .escn file that I can open in Godot and it contains most of the animation except the shape keys for the face.

**Comment by samvila** [comment]
One thing I notice, if I do the starch option to the shape keys in blender and then export it I get the animation keys for the shape keys in Godot but the name is wrong and that causes to fail the animation:

![wrong_names](https://user-images.githubusercontent.com/8132267/47960745-c95d9800-dfbc-11e8-8e16-b3bc9a665dcc.png)

Notice how the names that gets imported are: Armatureassistant_body while it should be assistant_body as far as I know.

**Comment by Jason0214** [comment]
These warnings are not related to shape keys, it is reported because the action for Armature is set to some mesh, just ignore them.  

Hmm, the name problem seems to be a broken in escn's structure..  I didn't think about any situation it could happen..  could you upload the escn file you exported? I may find some clue in it. 
By the way, what is a `starch` option for  shape key, sorry I haven't heard it before.

**Comment by samvila** [comment]
Here's the .escn file exported from Blender using the stash (sorry for the misspelled named). 


[sample5.escn.zip](https://github.com/godotengine/godot-blender-exporter/files/2545617/sample5.escn.zip)


Notice that the animationplayer has some blend shape keys but are pointing nowhere as far as I can tell.

**Comment by samvila** [comment]
I think I found the problem, the generated .escn writes the path as:

("Armatureassistant_body:blend_shapes/relax")

while the correct should be:

("Armature/assistant_body:blend_shapes/relax")

That fixes the problem, the problem now is that I have two animations on Godot, one for the body (bones) and another one for the shape keys and I can only play one at a time. I guess I might be able to merge both in Blender somehow and try to export it as a single animation but I might need to investigate how.

**Comment by Jason0214** [comment]
try the 'Scene Animation' option in the exporter, does this 'merge AnimationPlayer' solve your problem ? 
![image](https://user-images.githubusercontent.com/19422835/47961298-e480d580-dfc5-11e8-848f-61773cf85df7.png)

I will keep looking into the action path problem.

**Comment by samvila** [comment]
I tried the animation as actions with squash and that fixes my problem of having more than one animation on Godot, now all the animations are displayed in Godot as a single one which is what I was expected. Fixing manually the code to add the missing slash symbol fixes everything except one more thing I just notice... the shape keys on Godot are completely wrong from the ones in Blender, I think the way Godot is dealing with these shape keys values is wrong and this causes very weird animation.

Do you think I should open a new ticket only for that or should I continue on this one for this last problem?

**Comment by Jason0214** [comment]
I just go through the shape key playing, yeah, there seems some difference between godot and blender. So now there are two issues left:

1.  the action path one,  I just realized it is caused by I use `os.path` to handle action path, which behaves differently on windows.  I will open a thread for it myself

2. the shape key value one, I will do more check to investigate the problem and keep updating in this thread.

Really appreciate your patience and effort!

**Comment by dodgyville** [comment]
I've been trialing python 3.4's pathlib  over os.path and it seems to be better cross platform

**Comment by Jason0214** [comment]
Actually the capability of cross platform caused the issue, `os.path` gives backslash as separate symbol on windows, while for escn file, it should always separate with slash. Bad for me failing to realize it.

**Comment by samvila** [comment]
I think we are doing good progress here :-) , we figure out the problem with the slash symbol and not we are closer to have something quite functional. Assuming the slash symbol gets fixed I'm submitting another sample file:

[head_only.zip](https://github.com/godotengine/godot-blender-exporter/files/2545650/head_only.zip)

This one has two action with some shape keys, in this sample file we notice two issues:

- There's some problem with blending actions, in blender looks OK but on Godot is causes some weird animation
- The camera in Blender is set up to 35 but in Godot gets imported as 49.1 for whatever reason.

**Comment by Jason0214** [comment]
The camera is a none issue, blender default use focal length which the 35mm value you saw in your scene,  while godot use field of view angle.  If you switch the camera focal to FOV in blender, you will see it is exactly 49.1.

![image](https://user-images.githubusercontent.com/19422835/47968121-276c9880-e01a-11e8-8262-e61e731d6420.png)

---

The weird animation I noticed is the 'blink' one, in godot the mouth also moved as well as eye blinking, really weird..   I will look into it.

**Comment by Jason0214** [comment]
issues should be fixed after #107 and #108 merged in
